### PR TITLE
Add step learning rate scheduler

### PR DESCRIPTION
**Pull request by towerpark**
### Checklist

- [x] Confirmed that `run-checks all` script has been executed.
- [x] Made sure the book is up to date with changes in this PR.

The check script was executed with `DISABLE_WGPU=1`, because otherwise I would get an error similar to #1477, which is unrelated to this PR.

### Related Issues/PRs

#1198

### Changes

This PR adds the step learning rate scheduler, which is on the todo list in the issue above.

Initial learning rate and gamma are not checked when initializing a scheduler, because I think atypical values such as zero and negative values might be useful in some cases like [debugging](https://datascience.stackexchange.com/q/89518). Printing a warning is probably better but I'm not sure (I would be glad to hear your thoughts).

### Testing

* Added test cases for the scheduler
* Run `DISABLE_WGPU=1 bash ./run-checks.sh` and ensured that it exited with code 0
* Checked documentation by reading it in a browser
* Confirmed that the changes didn't affect the Books by searching LR in them and reading the contents

==========
Edit: Key mis-press (nothing changed).

### CODE REVIEW

File: crates/burn-core/src/lr_scheduler/step.rs [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+        }
+
+        Ok(StepLrScheduler {
+            init_lr: self.initial_lr,
+            step_size: self.step_size,
+            gamma: self.gamma,
+            iter_idx: -1,
+        })
+    }
+}
```
Comment (laggui) [inline crates/burn-core/src/lr_scheduler/step.rs:51]: Is there a particular reason why the initialization uses `Result` and returns and `Err` instead of assertions/panics for the value checks?

Just asking because the pattern is different from the other schedulers.

Also, regarding `gamma` and `init_lr`, I think a warnings would be a good idea as you suggested.

**Review by laggui — COMMENTED** [review]
Thanks for the addition!

Implementation looks good 🙂 

Just some minor comments

File: crates/burn-core/src/lr_scheduler/step.rs (hunk 1 above)
Comment (towerpark) [reply crates/burn-core/src/lr_scheduler/step.rs:51]: Thank you for your feedback.

I do have 2 reasons for using `Result`. One is for leaving callers the choice of how to handle the error. The other one is that returning `Result` seems acceptable since I find there are some other structs (not LR schedulers) doing so too. But neither of the reasons is a strong one and it does break the consistency, so if it's better to stick with the assertion way, I will fix that part.

As for the warning part, it seems that warnings should be printed by `log::warn!`, but I find a little issue with it. This method is a noop unless users explicitly create an application logger and call `.with_application_logger(None)` before building the learner. This is because the built-in logger is created in `LearnerBuilder::build`, which is unavailable when the LR scheduler is initialized.

I tried the MNIST example, however, and found that actually there were some log (`AdapterInfo` from cubecl_wgpu and `New execution plan 0` from burn_fusion) being discarded due to being logged at a stage too early, so letting users create the logger may not be that bad.

Comment (laggui) [reply crates/burn-core/src/lr_scheduler/step.rs:51]: I actually agree for `Result`, just that we would have to change the behavior of existing schedulers for consistency 😅 so for this PR addition I would keep it consistent with others, and possibly make the change to `Result` for all schedulers in a follow-up PR.

Regarding the warning, good catch! I'll think about what we should do, but the warning at the optimizer init level should be valid. If we're not logging because it hasn't been initialized, that's another issue to fix somewhere else I believe.

Comment (towerpark) [reply crates/burn-core/src/lr_scheduler/step.rs:51]: Yeah, one thing at a time is better. I will replace `Result` with an assertion and add warnings by `log::warn!` for initial LR and gamma, and leave the issues with returning value of `.init()` and logging to future PRs.

Comment (laggui) [reply crates/burn-core/src/lr_scheduler/step.rs:51]: Perfect. We should open issues for the follow-up works 👍

**Comment by towerpark** [comment]
I am sorry for the late update.

A fixup commit has been added to change the behavior of `StepLrSchedulerConfig::init`:
　* Make the method assert instead of returning `Result`
　* Add warning logs for atypical values of initial LR and gamma

Warning logs for initial LR and gamma are not getting test code, and were tested manually instead because there seems no straightforward way to do it. I have tried but couldn't come up with a solution. This is explained a little more in the source file.

**Review by laggui — APPROVED** [review]
No worries! Thanks for the addition 🙂
